plugins/ftdidmx/FtdiDmxThread.cpp

```diff
@@ -73,15 +99,164 @@ bool FtdiDmxThread::WriteDMX(const DmxBuffer &buffer) {
+        ola::NewSingleCallback(this,
+                               &FtdiDmxThread::DiscoveryComplete,
+                               callback));
+}
+
+void FtdiDmxThread::RunIncrementalDiscovery(rdm::RDMDiscoveryCallback *cb) {
+  m_discovery_agent.StartIncrementalDiscovery(
+        ola::NewSingleCallback(this,
+                               &FtdiDmxThread::DiscoveryComplete,
+                               cb));
+}
+
+/**
+ * Called when the discovery process finally completes
+ * @param callback the callback passed to StartFullDiscovery or
+ * StartIncrementalDiscovery that we should execute.
+ * @param status true if discovery worked, false otherwise
+ * @param uids the UIDSet of UIDs that were found.
+ */
+void FtdiDmxThread::DiscoveryComplete(ola::rdm::RDMDiscoveryCallback *callback,
+                                      bool status,
+                                      const ola::rdm::UIDSet &uids) {
+  if (status) {
+    OLA_DEBUG << "FTDI discovery complete: " << uids;
+  } else {
+    OLA_WARN << "FTDI discovery failed";
+  }
+  if (callback) {
+    callback->Run(uids);
+  }
+}
+
+/**
+ * @brief Method called to cleanup any outstanding callbacks
+ * @param state
+ *
+ * All callbacks except the RDMCallback lack a way of reporting an error state
+ * to the caller.
+ */
+void FtdiDmxThread::destroyPendindingCallback(ola::rdm::RDMStatusCode state) {
+  MuteDeviceCallback *thread_mute_callback = nullptr;
+  UnMuteDeviceCallback *thread_unmute_callback = nullptr;
+  BranchCallback *thread_branch_callback = nullptr;
+  ola::rdm::RDMCallback *thread_rdm_callback = nullptr;
+
+  destroyPendingRequest();
+
+  if (m_mute_complete != nullptr) {
+    thread_mute_callback = m_mute_complete;
+    m_mute_complete = nullptr;
+    thread_mute_callback->Run(false);
+  } else if (m_unmute_complete != nullptr) {
+    thread_unmute_callback = m_unmute_complete;
+    m_unmute_complete = nullptr;
+    thread_unmute_callback->Run();
+  } else if (m_branch_callback != nullptr) {
+    thread_branch_callback = m_branch_callback;
+    m_branch_callback = nullptr;
+    thread_branch_callback->Run(nullptr, 0);
+  } else if (m_rdm_callback != nullptr) {
+    thread_rdm_callback = m_rdm_callback;
+    m_rdm_callback = nullptr;
+    ola::rdm::RunRDMCallback(thread_rdm_callback, state);
+  }
+}
+
+void FtdiDmxThread::destroyPendingRequest() {
+  ola::rdm::RDMRequest *tmp_rdm_request = m_pending_request;
+
+  m_pending_request = nullptr;
+  delete tmp_rdm_request;
+}
+
+void FtdiDmxThread::MuteDevice(const ola::rdm::UID &target,
+                               MuteDeviceCallback *mute_complete) {
+  ola::thread::MutexLocker locker(&m_rdm_mutex);
+  if (m_pending_request == nullptr) {
+    OLA_INFO << "Muting device";
+    m_mute_complete = mute_complete;
+    m_pending_request = ola::rdm::NewMuteRequest(m_uid,
+                                                 target,
+                                                 m_transaction_number += 1);
```

Comment:
I think this is overridden in olad/OlaServerServiceImpl.cpp, so we should probably make the argument to NewMuteRequest optional...